Convert callback to `dyn` to improve compile times

Julian-Wollersberger opened the pull request:
To avoid monomorphizing `psm::on_stack::with_on_stack` 1500 times in rustc, this PR converts the callback passed to grow `_grow` to `dyn FnMut()`. The dynamic dispatch happens on the slow path.

This should improve rustc's compile time a bit, and surprisingly also improves it's runtime: [perf-results](https://perf.rust-lang.org/compare.html?start=4c1966f97e192d6282be935baa163fb58f9b8b27&end=345c9433914401e10ce9402a8b22f7e704f925f3) measured in rust-lang/rust#76680

This was discovered with the [cargo-llvm-lines](https://github.com/dtolnay/cargo-llvm-lines) tool.

Marwes commented on src/lib.rs:
```diff
@@ -59,11 +59,27 @@ pub fn maybe_grow<R, F: FnOnce() -> R>(red_zone: usize, stack_size: usize, callb
+    };
+
+    // _grow is only monomorphized once with a `dyn` argument.
+    _grow(stack_size, dyn_callback);
```
Just an observer to this, but personally I would change `_grow` to take `dyn FnMut()` directly, removing the possibility for it to be generic.

Avoids the risk of instantiating it multiple times.

nagisa replied:
Would make sense to me.

Julian-Wollersberger replied:
Done.

nagisa commented on src/lib.rs:
```diff
@@ -59,11 +59,27 @@ pub fn maybe_grow<R, F: FnOnce() -> R>(red_zone: usize, stack_size: usize, callb
 pub fn grow<R, F: FnOnce() -> R>(stack_size: usize, callback: F) -> R {
+    // Measuring with cargo-llvm-lines revealed that `psm::on_stack::with_on_stack`
+    // was monomorphized 1552 times in rustc and was responsible for 1.5% of
+    // rustc's total llvm IR lines. That takes time for LLVM to process.
```
This metric is probably very misleading, because `with_on_stack` is the base frame into which LLVM is able to inline the provided callback. `with_on_stack` is also super small itself (< dozen of instructions), so I have a hard time believing that by itself `with_on_stack` is a cause of the entirety of 1.5% of instructions.

With that in mind, making the callback dynamic only serves to prevent the inlining of the callback, and thus now all those closures end up having their own "category" or "line" in `cargo-llvm-lines` output. The improvement in perf/comptime still great, however.

Julian-Wollersberger replied:
I removed the numbers from the comment.

Yes, the actual functions that cargo-llvm-lines spits out are misleading, but they are a good to know where you can even _start looking_. And the total amount of llvm-lines decreased by 295 363 lines, so it's a big win. And yes, a separate entry for `stacker::grow::{{closure}}` appeared.

Julian-Wollersberger commented:
I addresses your reviews and also tried the [suggestion](https://github.com/rust-lang/rust/pull/76680#issuecomment-692958855) from @Marwes. It is a bit better to use a match instead of making a separate function for `enough_space`.

nagisa commented:
Released stacker 0.1.12 with this change.
